Broker startup fails when disk usage monitoring is disabled

**Describe the bug**

As [reported by the community](https://camunda-platform.slack.com/archives/C6WGNHV2A/p1665997293516809):

Starting a broker with disk usage monitoring disabled fails with an NPE while starting the command API.

**To Reproduce**

Set `ZEEBE_BROKER_DATA_DISKUSAGEMONITORINGENABLED` to false and start a broker

**Expected behavior**

Broker startup be successful, regardless of whether disk usage monitoring is enabled or not.

**Log/Stacktrace**

<!-- If possible add the full stacktrace or Zeebe log which contains the issue. -->

<details><summary>Full Stacktrace</summary>
 <p>

```
2022-10-17 10:47:11.339 [Broker-0-Startup] [Broker-0-zb-actors-1] WARN 
      io.camunda.zeebe.broker.system - Aborting startup process due to exception during step Command API
java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor.addDiskUsageListener(io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageListener)" because the return value of "io.camunda.zeebe.broker.bootstrap.BrokerStartupContext.getDiskSpaceUsageMonitor()" is null
	at io.camunda.zeebe.broker.bootstrap.CommandApiServiceStep.lambda$startCommandApiService$2(CommandApiServiceStep.java:91) ~[zeebe-broker-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.broker.bootstrap.AbstractBrokerStartupStep.forwardExceptions(AbstractBrokerStartupStep.java:71) ~[zeebe-broker-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.broker.bootstrap.AbstractBrokerStartupStep.lambda$proceed$3(AbstractBrokerStartupStep.java:87) ~[zeebe-broker-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
2022-10-17 10:47:11.355 [] [main] ERROR
      io.camunda.zeebe.broker.system - Failed to start broker 0!
java.util.concurrent.ExecutionException: Startup failed in the following steps: [Command API]. See suppressed exceptions for details.
	at io.camunda.zeebe.scheduler.future.CompletableActorFuture.get(CompletableActorFuture.java:142) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.future.CompletableActorFuture.get(CompletableActorFuture.java:109) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.FutureUtil.join(FutureUtil.java:21) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.future.CompletableActorFuture.join(CompletableActorFuture.java:198) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.broker.Broker.internalStart(Broker.java:101) ~[zeebe-broker-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.util.LogUtil.doWithMDC(LogUtil.java:23) ~[zeebe-util-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.broker.Broker.start(Broker.java:83) ~[zeebe-broker-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.broker.StandaloneBroker.run(StandaloneBroker.java:92) ~[camunda-zeebe-8.1.0.jar:8.1.0]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:771) ~[spring-boot-2.7.4.jar:2.7.4]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:755) ~[spring-boot-2.7.4.jar:2.7.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-2.7.4.jar:2.7.4]
	at io.camunda.zeebe.broker.StandaloneBroker.main(StandaloneBroker.java:82) ~[camunda-zeebe-8.1.0.jar:8.1.0]
Caused by: io.camunda.zeebe.scheduler.startup.StartupProcessException: Startup failed in the following steps: [Command API]. See suppressed exceptions for details.
	at io.camunda.zeebe.scheduler.startup.StartupProcess.aggregateExceptionsSynchronized(StartupProcess.java:282) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.startup.StartupProcess.completeStartupFutureExceptionallySynchronized(StartupProcess.java:183) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.startup.StartupProcess.lambda$proceedWithStartupSynchronized$3(StartupProcess.java:167) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	Suppressed: io.camunda.zeebe.scheduler.startup.StartupProcessStepException: Bootstrap step Command API failed
		at io.camunda.zeebe.scheduler.startup.StartupProcess.completeStartupFutureExceptionallySynchronized(StartupProcess.java:185) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.startup.StartupProcess.lambda$proceedWithStartupSynchronized$3(StartupProcess.java:167) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
	Caused by: java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor.addDiskUsageListener(io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageListener)" because the return value of "io.camunda.zeebe.broker.bootstrap.BrokerStartupContext.getDiskSpaceUsageMonitor()" is null
		at io.camunda.zeebe.broker.bootstrap.CommandApiServiceStep.lambda$startCommandApiService$2(CommandApiServiceStep.java:91) ~[zeebe-broker-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.broker.bootstrap.AbstractBrokerStartupStep.forwardExceptions(AbstractBrokerStartupStep.java:71) ~[zeebe-broker-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.broker.bootstrap.AbstractBrokerStartupStep.lambda$proceed$3(AbstractBrokerStartupStep.java:87) ~[zeebe-broker-8.1.0.jar:8.1.0]
		at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) ~[zeebe-scheduler-8.1.0.jar:8.1.0]
		... 6 more```

</p>
</details>

**Environment:**
- OS: <!-- [e.g. Linux] -->
- Zeebe Version: 8.1.0
- Configuration: `ZEEBE_BROKER_DATA_DISKUSAGEMONITORINGENABLED=false`

--- a/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/BrokerStartupProcess.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/BrokerStartupProcess.java
@@ -48,10 +48,9 @@ public final class BrokerStartupProcess {
 
     result.add(new ClusterServicesStep());
 
-    if (config.getData().isDiskUsageMonitoringEnabled()) {
-      // must be executed before any disk space usage listeners are registered
-      result.add(new DiskSpaceUsageMonitorStep());
-    }
+    // must be executed before any disk space usage listeners are registered
+    result.add(new DiskSpaceUsageMonitorStep());
+
     result.add(new MonitoringServerStep());
     result.add(new BrokerAdminServiceStep());
 

--- a/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
@@ -13,6 +13,7 @@ import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor;
 import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitorActor;
 import io.camunda.zeebe.scheduler.ConcurrencyControl;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
+import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
 import io.camunda.zeebe.util.FileUtil;
 import java.io.IOException;
 import java.nio.file.Paths;
@@ -43,13 +44,31 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
       final ConcurrencyControl concurrencyControl,
       final ActorFuture<BrokerStartupContext> shutdownFuture) {
 
-    final var diskSpaceUsageMonitor = brokerShutdownContext.getDiskSpaceUsageMonitor();
-    if (diskSpaceUsageMonitor instanceof DiskSpaceUsageMonitorActor actor) {
-      stopDiskUsageMonitorActor(brokerShutdownContext, concurrencyControl, shutdownFuture, actor);
-    } else {
-      brokerShutdownContext.setDiskSpaceUsageMonitor(null);
+    if (brokerShutdownContext.getDiskSpaceUsageMonitor() == null) {
       shutdownFuture.complete(brokerShutdownContext);
+      return;
     }
+
+    concurrencyControl.runOnCompletion(
+        brokerShutdownContext.getDiskSpaceUsageMonitor().closeAsync(),
+        (ok, error) -> {
+          if (error != null) {
+            shutdownFuture.completeExceptionally(error);
+            return;
+          }
+
+          forwardExceptions(
+              () ->
+                  concurrencyControl.run(
+                      () ->
+                          forwardExceptions(
+                              () -> {
+                                brokerShutdownContext.setDiskSpaceUsageMonitor(null);
+                                shutdownFuture.complete(brokerShutdownContext);
+                              },
+                              shutdownFuture)),
+              shutdownFuture);
+        });
   }
 
   private static void startDiskUsageMonitorActor(
@@ -83,34 +102,6 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
             });
   }
 
-  private void stopDiskUsageMonitorActor(
-      final BrokerStartupContext brokerShutdownContext,
-      final ConcurrencyControl concurrencyControl,
-      final ActorFuture<BrokerStartupContext> shutdownFuture,
-      final DiskSpaceUsageMonitorActor actor) {
-    final var closeFuture = actor.closeAsync();
-    concurrencyControl.runOnCompletion(
-        closeFuture,
-        (ok, error) -> {
-          if (error != null) {
-            shutdownFuture.completeExceptionally(error);
-            return;
-          }
-
-          forwardExceptions(
-              () ->
-                  concurrencyControl.run(
-                      () ->
-                          forwardExceptions(
-                              () -> {
-                                brokerShutdownContext.setDiskSpaceUsageMonitor(null);
-                                shutdownFuture.complete(brokerShutdownContext);
-                              },
-                              shutdownFuture)),
-              shutdownFuture);
-        });
-  }
-
   @Override
   public String getName() {
     return "Disk Space Usage Monitor";
@@ -126,5 +117,10 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
 
     @Override
     public void setFreeDiskSpaceSupplier(final LongSupplier freeDiskSpaceSupplier) {}
+
+    @Override
+    public ActorFuture<Void> closeAsync() {
+      return CompletableActorFuture.completed(null);
+    }
   }
 }

--- a/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
@@ -7,7 +7,7 @@
  */
 package io.camunda.zeebe.broker.bootstrap;
 
-import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor;
+import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitorActor;
 import io.camunda.zeebe.scheduler.ConcurrencyControl;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.util.FileUtil;
@@ -30,7 +30,7 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
       return;
     }
 
-    final var diskSpaceUsageMonitor = new DiskSpaceUsageMonitor(data);
+    final var diskSpaceUsageMonitor = new DiskSpaceUsageMonitorActor(data);
 
     final var actorStartFuture =
         brokerStartupContext.getActorSchedulingService().submitActor(diskSpaceUsageMonitor);
@@ -57,9 +57,8 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
       final ActorFuture<BrokerStartupContext> shutdownFuture) {
 
     final var diskSpaceUsageMonitor = brokerShutdownContext.getDiskSpaceUsageMonitor();
-
-    if (diskSpaceUsageMonitor != null) {
-      final var closeFuture = diskSpaceUsageMonitor.closeAsync();
+    if (diskSpaceUsageMonitor instanceof DiskSpaceUsageMonitorActor actor) {
+      final var closeFuture = actor.closeAsync();
       concurrencyControl.runOnCompletion(
           closeFuture,
           (ok, error) -> {

--- a/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
@@ -7,6 +7,7 @@
  */
 package io.camunda.zeebe.broker.bootstrap;
 
+import io.camunda.zeebe.broker.system.configuration.DataCfg;
 import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitorActor;
 import io.camunda.zeebe.scheduler.ConcurrencyControl;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
@@ -23,6 +24,28 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
       final ActorFuture<BrokerStartupContext> startupFuture) {
 
     final var data = brokerStartupContext.getBrokerConfiguration().getData();
+
+    startDiskUsageMonitorActor(brokerStartupContext, startupFuture, data);
+  }
+
+  @Override
+  void shutdownInternal(
+      final BrokerStartupContext brokerShutdownContext,
+      final ConcurrencyControl concurrencyControl,
+      final ActorFuture<BrokerStartupContext> shutdownFuture) {
+
+    final var diskSpaceUsageMonitor = brokerShutdownContext.getDiskSpaceUsageMonitor();
+    if (diskSpaceUsageMonitor instanceof DiskSpaceUsageMonitorActor actor) {
+      stopDiskUsageMonitorActor(brokerShutdownContext, concurrencyControl, shutdownFuture, actor);
+    } else {
+      shutdownFuture.complete(brokerShutdownContext);
+    }
+  }
+
+  private static void startDiskUsageMonitorActor(
+      final BrokerStartupContext brokerStartupContext,
+      final ActorFuture<BrokerStartupContext> startupFuture,
+      final DataCfg data) {
     try {
       FileUtil.ensureDirectoryExists(Paths.get(data.getDirectory()));
     } catch (final IOException e) {
@@ -50,38 +73,32 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
             });
   }
 
-  @Override
-  void shutdownInternal(
+  private void stopDiskUsageMonitorActor(
       final BrokerStartupContext brokerShutdownContext,
       final ConcurrencyControl concurrencyControl,
-      final ActorFuture<BrokerStartupContext> shutdownFuture) {
+      final ActorFuture<BrokerStartupContext> shutdownFuture,
+      final DiskSpaceUsageMonitorActor actor) {
+    final var closeFuture = actor.closeAsync();
+    concurrencyControl.runOnCompletion(
+        closeFuture,
+        (ok, error) -> {
+          if (error != null) {
+            shutdownFuture.completeExceptionally(error);
+            return;
+          }
 
-    final var diskSpaceUsageMonitor = brokerShutdownContext.getDiskSpaceUsageMonitor();
-    if (diskSpaceUsageMonitor instanceof DiskSpaceUsageMonitorActor actor) {
-      final var closeFuture = actor.closeAsync();
-      concurrencyControl.runOnCompletion(
-          closeFuture,
-          (ok, error) -> {
-            if (error != null) {
-              shutdownFuture.completeExceptionally(error);
-              return;
-            }
-
-            forwardExceptions(
-                () ->
-                    concurrencyControl.run(
-                        () ->
-                            forwardExceptions(
-                                () -> {
-                                  brokerShutdownContext.setDiskSpaceUsageMonitor(null);
-                                  shutdownFuture.complete(brokerShutdownContext);
-                                },
-                                shutdownFuture)),
-                shutdownFuture);
-          });
-    } else {
-      shutdownFuture.complete(brokerShutdownContext);
-    }
+          forwardExceptions(
+              () ->
+                  concurrencyControl.run(
+                      () ->
+                          forwardExceptions(
+                              () -> {
+                                brokerShutdownContext.setDiskSpaceUsageMonitor(null);
+                                shutdownFuture.complete(brokerShutdownContext);
+                              },
+                              shutdownFuture)),
+              shutdownFuture);
+        });
   }
 
   @Override

--- a/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStep.java
@@ -8,12 +8,15 @@
 package io.camunda.zeebe.broker.bootstrap;
 
 import io.camunda.zeebe.broker.system.configuration.DataCfg;
+import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageListener;
+import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor;
 import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitorActor;
 import io.camunda.zeebe.scheduler.ConcurrencyControl;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.util.FileUtil;
 import java.io.IOException;
 import java.nio.file.Paths;
+import java.util.function.LongSupplier;
 
 class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
 
@@ -24,6 +27,12 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
       final ActorFuture<BrokerStartupContext> startupFuture) {
 
     final var data = brokerStartupContext.getBrokerConfiguration().getData();
+
+    if (!data.isDiskUsageMonitoringEnabled()) {
+      brokerStartupContext.setDiskSpaceUsageMonitor(new DisabledDiskUsageMonitor());
+      startupFuture.complete(brokerStartupContext);
+      return;
+    }
 
     startDiskUsageMonitorActor(brokerStartupContext, startupFuture, data);
   }
@@ -38,6 +47,7 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
     if (diskSpaceUsageMonitor instanceof DiskSpaceUsageMonitorActor actor) {
       stopDiskUsageMonitorActor(brokerShutdownContext, concurrencyControl, shutdownFuture, actor);
     } else {
+      brokerShutdownContext.setDiskSpaceUsageMonitor(null);
       shutdownFuture.complete(brokerShutdownContext);
     }
   }
@@ -104,5 +114,17 @@ class DiskSpaceUsageMonitorStep extends AbstractBrokerStartupStep {
   @Override
   public String getName() {
     return "Disk Space Usage Monitor";
+  }
+
+  private static final class DisabledDiskUsageMonitor implements DiskSpaceUsageMonitor {
+
+    @Override
+    public void addDiskUsageListener(final DiskSpaceUsageListener listener) {}
+
+    @Override
+    public void removeDiskUsageListener(final DiskSpaceUsageListener listener) {}
+
+    @Override
+    public void setFreeDiskSpaceSupplier(final LongSupplier freeDiskSpaceSupplier) {}
   }
 }

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitor.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitor.java
@@ -7,78 +7,14 @@
  */
 package io.camunda.zeebe.broker.system.monitoring;
 
-import static io.camunda.zeebe.broker.Broker.LOG;
-
-import io.camunda.zeebe.broker.system.configuration.DataCfg;
-import io.camunda.zeebe.scheduler.Actor;
-import java.io.File;
-import java.io.IOException;
-import java.io.UncheckedIOException;
-import java.time.Duration;
-import java.util.HashSet;
-import java.util.Set;
 import java.util.function.LongSupplier;
 
-public class DiskSpaceUsageMonitor extends Actor {
+public interface DiskSpaceUsageMonitor {
 
-  private final Set<DiskSpaceUsageListener> diskSpaceUsageListeners = new HashSet<>();
-  private boolean currentDiskAvailableStatus = true;
-  private LongSupplier freeDiskSpaceSupplier;
-  private final Duration monitoringDelay;
-  private final long minFreeDiskSpaceRequired;
+  void addDiskUsageListener(DiskSpaceUsageListener listener);
 
-  public DiskSpaceUsageMonitor(final DataCfg dataCfg) {
-    monitoringDelay = dataCfg.getDiskUsageMonitoringInterval();
-    minFreeDiskSpaceRequired = dataCfg.getFreeDiskSpaceCommandWatermark();
-    final var directory = new File(dataCfg.getDirectory());
-
-    if (!directory.exists()) {
-      throw new UncheckedIOException(new IOException("Folder '" + directory + "' does not exist."));
-    }
-    freeDiskSpaceSupplier = directory::getUsableSpace;
-  }
-
-  @Override
-  protected void onActorStarted() {
-    checkDiskUsageAndNotifyListeners();
-    actor.runAtFixedRate(monitoringDelay, this::checkDiskUsageAndNotifyListeners);
-  }
-
-  private void checkDiskUsageAndNotifyListeners() {
-    final long freeDiskSpaceAvailable = freeDiskSpaceSupplier.getAsLong();
-    final boolean previousStatus = currentDiskAvailableStatus;
-    currentDiskAvailableStatus = freeDiskSpaceAvailable >= minFreeDiskSpaceRequired;
-    if (currentDiskAvailableStatus != previousStatus) {
-      if (!currentDiskAvailableStatus) {
-        LOG.warn(
-            "Out of disk space. Current available {} bytes. Minimum needed {} bytes.",
-            freeDiskSpaceAvailable,
-            minFreeDiskSpaceRequired);
-        diskSpaceUsageListeners.forEach(DiskSpaceUsageListener::onDiskSpaceNotAvailable);
-      } else {
-        LOG.info("Disk space available again. Current available {} bytes", freeDiskSpaceAvailable);
-        diskSpaceUsageListeners.forEach(DiskSpaceUsageListener::onDiskSpaceAvailable);
-      }
-    }
-  }
-
-  public void addDiskUsageListener(final DiskSpaceUsageListener listener) {
-    actor.call(
-        () -> {
-          diskSpaceUsageListeners.add(listener);
-          // Listeners always assumes diskspace is available on start. So notify them if it is not.
-          if (!currentDiskAvailableStatus) {
-            listener.onDiskSpaceNotAvailable();
-          }
-        });
-  }
-
-  public void removeDiskUsageListener(final DiskSpaceUsageListener listener) {
-    actor.call(() -> diskSpaceUsageListeners.remove(listener));
-  }
+  void removeDiskUsageListener(DiskSpaceUsageListener listener);
 
   // Used only for testing
-  public void setFreeDiskSpaceSupplier(final LongSupplier freeDiskSpaceSupplier) {
-    this.freeDiskSpaceSupplier = freeDiskSpaceSupplier;
-  }
+  void setFreeDiskSpaceSupplier(LongSupplier freeDiskSpaceSupplier);
 }

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitor.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitor.java
@@ -7,9 +7,10 @@
  */
 package io.camunda.zeebe.broker.system.monitoring;
 
+import io.camunda.zeebe.scheduler.AsyncClosable;
 import java.util.function.LongSupplier;
 
-public interface DiskSpaceUsageMonitor {
+public interface DiskSpaceUsageMonitor extends AsyncClosable {
 
   void addDiskUsageListener(DiskSpaceUsageListener listener);
 

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitorActor.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitorActor.java
@@ -1,0 +1,87 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.broker.system.monitoring;
+
+import static io.camunda.zeebe.broker.Broker.LOG;
+
+import io.camunda.zeebe.broker.system.configuration.DataCfg;
+import io.camunda.zeebe.scheduler.Actor;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.LongSupplier;
+
+public class DiskSpaceUsageMonitorActor extends Actor implements DiskSpaceUsageMonitor {
+
+  private final Set<DiskSpaceUsageListener> diskSpaceUsageListeners = new HashSet<>();
+  private boolean currentDiskAvailableStatus = true;
+  private LongSupplier freeDiskSpaceSupplier;
+  private final Duration monitoringDelay;
+  private final long minFreeDiskSpaceRequired;
+
+  public DiskSpaceUsageMonitorActor(final DataCfg dataCfg) {
+    monitoringDelay = dataCfg.getDiskUsageMonitoringInterval();
+    minFreeDiskSpaceRequired = dataCfg.getFreeDiskSpaceCommandWatermark();
+    final var directory = new File(dataCfg.getDirectory());
+
+    if (!directory.exists()) {
+      throw new UncheckedIOException(new IOException("Folder '" + directory + "' does not exist."));
+    }
+    freeDiskSpaceSupplier = directory::getUsableSpace;
+  }
+
+  @Override
+  protected void onActorStarted() {
+    checkDiskUsageAndNotifyListeners();
+    actor.runAtFixedRate(monitoringDelay, this::checkDiskUsageAndNotifyListeners);
+  }
+
+  private void checkDiskUsageAndNotifyListeners() {
+    final long freeDiskSpaceAvailable = freeDiskSpaceSupplier.getAsLong();
+    final boolean previousStatus = currentDiskAvailableStatus;
+    currentDiskAvailableStatus = freeDiskSpaceAvailable >= minFreeDiskSpaceRequired;
+    if (currentDiskAvailableStatus != previousStatus) {
+      if (!currentDiskAvailableStatus) {
+        LOG.warn(
+            "Out of disk space. Current available {} bytes. Minimum needed {} bytes.",
+            freeDiskSpaceAvailable,
+            minFreeDiskSpaceRequired);
+        diskSpaceUsageListeners.forEach(DiskSpaceUsageListener::onDiskSpaceNotAvailable);
+      } else {
+        LOG.info("Disk space available again. Current available {} bytes", freeDiskSpaceAvailable);
+        diskSpaceUsageListeners.forEach(DiskSpaceUsageListener::onDiskSpaceAvailable);
+      }
+    }
+  }
+
+  @Override
+  public void addDiskUsageListener(final DiskSpaceUsageListener listener) {
+    actor.call(
+        () -> {
+          diskSpaceUsageListeners.add(listener);
+          // Listeners always assumes diskspace is available on start. So notify them if it is not.
+          if (!currentDiskAvailableStatus) {
+            listener.onDiskSpaceNotAvailable();
+          }
+        });
+  }
+
+  @Override
+  public void removeDiskUsageListener(final DiskSpaceUsageListener listener) {
+    actor.call(() -> diskSpaceUsageListeners.remove(listener));
+  }
+
+  // Used only for testing
+  @Override
+  public void setFreeDiskSpaceSupplier(final LongSupplier freeDiskSpaceSupplier) {
+    this.freeDiskSpaceSupplier = freeDiskSpaceSupplier;
+  }
+}

--- a/broker/src/test/java/io/camunda/zeebe/broker/bootstrap/CommandApiServiceStepTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/bootstrap/CommandApiServiceStepTest.java
@@ -20,7 +20,7 @@ import io.camunda.zeebe.broker.clustering.ClusterServicesImpl;
 import io.camunda.zeebe.broker.exporter.repo.ExporterRepository;
 import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
 import io.camunda.zeebe.broker.system.monitoring.BrokerHealthCheckService;
-import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor;
+import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitorActor;
 import io.camunda.zeebe.broker.transport.commandapi.CommandApiServiceImpl;
 import io.camunda.zeebe.protocol.impl.encoding.BrokerInfo;
 import io.camunda.zeebe.scheduler.ActorScheduler;
@@ -68,7 +68,7 @@ class CommandApiServiceStepTest {
             mock(ClusterServicesImpl.class, RETURNS_DEEP_STUBS),
             Collections.emptyList());
     testBrokerStartupContext.setConcurrencyControl(CONCURRENCY_CONTROL);
-    testBrokerStartupContext.setDiskSpaceUsageMonitor(mock(DiskSpaceUsageMonitor.class));
+    testBrokerStartupContext.setDiskSpaceUsageMonitor(mock(DiskSpaceUsageMonitorActor.class));
     testBrokerStartupContext.setGatewayBrokerTransport(mock(AtomixServerTransport.class));
   }
 
@@ -135,7 +135,7 @@ class CommandApiServiceStepTest {
     @Test
     void shouldAddCommandApiServiceAsDiskSpaceUsageListener() {
       // given
-      final var mockDiskSpaceUsageMonitor = mock(DiskSpaceUsageMonitor.class);
+      final var mockDiskSpaceUsageMonitor = mock(DiskSpaceUsageMonitorActor.class);
       testBrokerStartupContext.setDiskSpaceUsageMonitor(mockDiskSpaceUsageMonitor);
 
       // when
@@ -171,7 +171,7 @@ class CommandApiServiceStepTest {
       testBrokerStartupContext.setGatewayBrokerTransport(mockAtomixServerTransport);
       testBrokerStartupContext.setCommandApiService(mockCommandApiService);
       testBrokerStartupContext.addPartitionListener(mockCommandApiService);
-      testBrokerStartupContext.setDiskSpaceUsageMonitor(mock(DiskSpaceUsageMonitor.class));
+      testBrokerStartupContext.setDiskSpaceUsageMonitor(mock(DiskSpaceUsageMonitorActor.class));
       testBrokerStartupContext
           .getDiskSpaceUsageMonitor()
           .addDiskUsageListener(mockCommandApiService);
@@ -182,7 +182,7 @@ class CommandApiServiceStepTest {
     @Test
     void shouldRemoveCommandApiFromDiskSpaceUsageListenerList() {
       // given
-      final var mockDiskSpaceUsageMonitor = mock(DiskSpaceUsageMonitor.class);
+      final var mockDiskSpaceUsageMonitor = mock(DiskSpaceUsageMonitorActor.class);
       testBrokerStartupContext.setDiskSpaceUsageMonitor(mockDiskSpaceUsageMonitor);
 
       // when

--- a/broker/src/test/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStepTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/bootstrap/DiskSpaceUsageMonitorStepTest.java
@@ -15,7 +15,7 @@ import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
-import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitor;
+import io.camunda.zeebe.broker.system.monitoring.DiskSpaceUsageMonitorActor;
 import io.camunda.zeebe.scheduler.ActorSchedulingService;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.scheduler.testing.TestConcurrencyControl;
@@ -31,7 +31,7 @@ class DiskSpaceUsageMonitorStepTest {
 
   private BrokerStartupContext mockBrokerStartupContext;
   private ActorSchedulingService mockActorSchedulingService;
-  private DiskSpaceUsageMonitor mockDiskSpaceUsageMonitor;
+  private DiskSpaceUsageMonitorActor mockDiskSpaceUsageMonitor;
 
   private ActorFuture<BrokerStartupContext> future;
 
@@ -41,7 +41,7 @@ class DiskSpaceUsageMonitorStepTest {
   void setUp() {
     mockBrokerStartupContext = mock(BrokerStartupContext.class);
     mockActorSchedulingService = mock(ActorSchedulingService.class);
-    mockDiskSpaceUsageMonitor = mock(DiskSpaceUsageMonitor.class);
+    mockDiskSpaceUsageMonitor = mock(DiskSpaceUsageMonitorActor.class);
 
     when(mockBrokerStartupContext.getBrokerConfiguration()).thenReturn(TEST_BROKER_CONFIG);
     when(mockBrokerStartupContext.getConcurrencyControl()).thenReturn(CONCURRENCY_CONTROL);
@@ -73,7 +73,7 @@ class DiskSpaceUsageMonitorStepTest {
     sut.startupInternal(mockBrokerStartupContext, CONCURRENCY_CONTROL, future);
 
     // then
-    final var argumentCaptor = ArgumentCaptor.forClass(DiskSpaceUsageMonitor.class);
+    final var argumentCaptor = ArgumentCaptor.forClass(DiskSpaceUsageMonitorActor.class);
     verify(mockBrokerStartupContext).setDiskSpaceUsageMonitor(argumentCaptor.capture());
     verify(mockActorSchedulingService).submitActor(argumentCaptor.getValue());
   }

--- a/broker/src/test/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitorTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/system/monitoring/DiskSpaceUsageMonitorTest.java
@@ -22,7 +22,7 @@ public class DiskSpaceUsageMonitorTest {
     dataCfg.setDirectory("something-that-does-not-exist");
 
     // when + then
-    assertThatThrownBy(() -> new DiskSpaceUsageMonitor(dataCfg))
+    assertThatThrownBy(() -> new DiskSpaceUsageMonitorActor(dataCfg))
         .isInstanceOf(UncheckedIOException.class)
         .hasMessage("java.io.IOException: Folder 'something-that-does-not-exist' does not exist.");
   }

